pkg/dynamicinstrumentation/diconfig/file_config_manager.go

```diff
@@ -3,228 +3,45 @@
-// which contains probe configurations
-type FileWatchingConfigManager struct {
-	sync.Mutex
-	configTracker *configTracker
-	procTracker   *proctracker.ProcessTracker
-
-	callback configUpdateCallback
-	configs  configsByService
-	state    ditypes.DIProcs
-}
-
-type fileConfigCallback func(configsByService)
-
-type configsByService = map[ditypes.ServiceName]map[ditypes.ProbeID]rcConfig
-
-// NewFileConfigManager creates a FileWatchingConfigManager set up to track
-// the specified file.
-func NewFileConfigManager(configFile string) (*FileWatchingConfigManager, error) {
-	cm := &FileWatchingConfigManager{
-		callback: applyConfigUpdate,
-	}
-
-	cm.procTracker = proctracker.NewProcessTracker(cm.updateProcessInfo)
-	err := cm.procTracker.Start()
-	if err != nil {
-		return nil, err
-	}
-
-	cm.configTracker = newFileWatchingConfigTracker(configFile, cm.updateServiceConfigs)
-	err = cm.configTracker.Start()
+func NewFileConfigManager(configFile string) (*ReaderConfigManager, error) {
+	cm, err := NewReaderConfigManager()
 	if err != nil {
 		return nil, err
 	}
-	return cm, nil
-}
 
-// GetProcInfos returns the state of the FileWatchingConfigManager
-func (cm *FileWatchingConfigManager) GetProcInfos() ditypes.DIProcs {
-	return cm.state
-}
-
-// Stop closes the config and proc trackers used by the FileWatchingConfigManager
-func (cm *FileWatchingConfigManager) Stop() {
-	cm.configTracker.Stop()
-	cm.procTracker.Stop()
-}
-
-func newFileWatchingConfigTracker(configFile string, onConfigUpdate fileConfigCallback) *configTracker {
-	ct := configTracker{
-		ConfigPath:     configFile,
-		configCallback: onConfigUpdate,
-		stopChannel:    make(chan bool),
-	}
-
-	return &ct
-}
-
-// correlate this new configuration with a running service,
-// and operate on the new global state of services/configs
-// via cm.callback
-func (cm *FileWatchingConfigManager) updateServiceConfigs(configs configsByService) {
-	log.Info("Updating config from file:", configs)
-	cm.configs = configs
-	err := cm.update()
-	if err != nil {
-		log.Info(err)
-	}
-}
-
-func (cm *FileWatchingConfigManager) updateProcessInfo(procs ditypes.DIProcs) {
-	cm.Lock()
-	defer cm.Unlock()
-	log.Info("Updating procs", procs)
-	cm.configTracker.UpdateProcesses(procs)
-	err := cm.update()
-	if err != nil {
-		log.Info(err)
-	}
-}
-
-type configTracker struct {
-	Processes      map[ditypes.PID]*ditypes.ProcessInfo
-	ConfigPath     string
-	configCallback fileConfigCallback
-	stopChannel    chan bool
-}
-
-func (ct *configTracker) Start() error {
-	fw := util.NewFileWatcher(ct.ConfigPath)
+	fw := util.NewFileWatcher(configFile)
 	updateChan, err := fw.Watch()
 	if err != nil {
-		return fmt.Errorf("failed to watch config file %s: %s", ct.ConfigPath, err)
+		return nil, fmt.Errorf("failed to watch config file %s: %s", configFile, err)
 	}
 
-	go func(updateChan <-chan []byte) {
-	configUpdateLoop:
+	c := make(chan os.Signal, 1)
+	signal.Notify(c, os.Interrupt)
```

Comment:
It is a bit strange to handle the signal this deep. Also, a signal isn't the only way you want graceful termination. Within a test, you will want to stop everything gracefully, but there is no signal at test end.

A couple ways you could solve this:
- return a function that closes the channel, which will exit the goroutine. Call the function in the parent `Close` function.
- create a wrapping type here that provides a `Stop/Close` function, and call that in the parent.